File: py/tools/parity_roundtrip.py
"""parity_roundtrip — read a crtx v0.1 envelope JSON file at argv[1],
parse via ``stem.parse_envelope``, re-serialize via
``stem.serialize_envelope``, write the result to stdout.

Used by ``tools/parity/runner.sh``. Run from the repo root as::

    cd py && PYTHONPATH=src python3 tools/parity_roundtrip.py <path>

Or via uv::

    cd py && uv run python tools/parity_roundtrip.py <path>
"""
from __future__ import annotations

import sys
from pathlib import Path

from stem import parse_envelope, serialize_envelope


def main(argv: list[str]) -> int:
    if len(argv) != 2:
        print("usage: parity_roundtrip.py <envelope.json>", file=sys.stderr)
        return 2

    fixture = Path(argv[1])
    try:
        text = fixture.read_text(encoding="utf-8")
    except OSError as err:
        print(f"parity-roundtrip: read {fixture}: {err}", file=sys.stderr)
        return 1

    try:
        env = parse_envelope(text)
    except Exception as err:  # noqa: BLE001 — surface every parse failure
        print(f"parity-roundtrip: parse: {err}", file=sys.stderr)
        return 1

    try:
        out = serialize_envelope(env)
    except Exception as err:  # noqa: BLE001
        print(f"parity-roundtrip: serialize: {err}", file=sys.stderr)
        return 1

    sys.stdout.write(out)
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: php/tools/parity-roundtrip.php
<?php

declare(strict_types=1);

/**
 * parity-roundtrip — read a crtx v0.1 envelope JSON file at argv[1],
 * parse via HopTop\Stem\parseEnvelope, re-serialize via
 * HopTop\Stem\serializeEnvelope, write the result to stdout.
 *
 * Used by tools/parity/runner.sh. Run from php/ as:
 *
 *   php tools/parity-roundtrip.php <path>
 */

require __DIR__ . '/../vendor/autoload.php';

use function HopTop\Stem\parseEnvelope;
use function HopTop\Stem\serializeEnvelope;

if ($argc !== 2) {
    fwrite(STDERR, "usage: parity-roundtrip <envelope.json>\n");
    exit(2);
}

$path = $argv[1];
$text = @file_get_contents($path);
if ($text === false) {
    fwrite(STDERR, sprintf("parity-roundtrip: read %s: %s\n", $path, error_get_last()['message'] ?? 'unknown error'));
    exit(1);
}

try {
    $env = parseEnvelope($text);
} catch (\Throwable $e) {
    fwrite(STDERR, sprintf("parity-roundtrip: parse: %s\n", $e->getMessage()));
    exit(1);
}

try {
    $out = serializeEnvelope($env);
} catch (\Throwable $e) {
    fwrite(STDERR, sprintf("parity-roundtrip: serialize: %s\n", $e->getMessage()));
    exit(1);
}

fwrite(STDOUT, $out);
